src/math_tools/curl.hpp

```diff
@@ -0,0 +1,65 @@
+// SPDX-License-Identifier: MIT
+#pragma once
+
+#include "ddc_alias_inline_functions.hpp"
+#include "ddc_aliases.hpp"
+#include "gradient.hpp"
+#include "metric_tensor_evaluator.hpp"
+#include "static_tensors.hpp"
+#include "vector_field.hpp"
+
+/**
+ * @brief A class which implements a curl operator
+ * The implemented equation is:
+ * @f$ \nabla \cross \mathbf{F} @f$
+ * @f$ \nabla \cross \mathbf{F} = {\cal J}_{\rm x}^{-1}\epsilon^{klm}\partial_{x^l} F_m @f$
+ * with @f$ {\cal J}_{\rm x} @f$ the jacobian of the system,
+ * @f$ F_m @f$ the covariant components of F and @f$\epsilon^{klm} @f$ the Levi-Civita symbol.
+ * @tparam Mapping3D A type representing a mapping in 3 dimensions.
+ * @tparam MappingCoord The type of the coordinate that will be used to evaluate the
+ *                  mapping. This coordinate is used to calculate the determinant of
+ *                  the Jacobian and the metric tensor. It is almost always equal to
+ *                  the argument type of the 3D mapping but axi-symmetry may make it
+ *                  useful to evaluate the mapping on a coordinate with a reduced
+ *                  dimensionality.
+ */
+template <class Mapping3D, class MappingCoord = typename Mapping3D::CoordArg>
+class Curl
+{
+    static_assert(is_mapping_v<Mapping3D>);
+
+    using BasisSpatial = ddc::to_type_seq_t<typename Mapping3D::CoordArg>;
+    using CovBasisSpatial = get_covariant_dims_t<BasisSpatial>;
+
+private:
+    Mapping3D m_mapping;
+    MetricTensorEvaluator<Mapping3D, MappingCoord> m_metric_tensor;
+
+public:
+    /**
+     * @brief Build a curl operator.
+     * @param mapping The mapping describing the system of coordinates on which the
+     *              expression is calculated.
+     */
+    explicit Curl(Mapping3D const& mapping) : m_mapping(mapping), m_metric_tensor(mapping) {}
+
+    /**
+     * @brief Compute a curl at a given coordinate, from
+     * the partial derivatives of the vector field f
+     *
+     * @param[in] partial_derivatives_f A tensor containing the partial derivatives
+     * of the vector field f expressed at the given coordinate.
+     * @param[in] coord The coordinate where the calculation is carried out.
+     */
+    template <
+            class TensorType,
+            class = std::enable_if_t<is_tensor_type_v<TensorType> && TensorType::rank() == 2>>
+    KOKKOS_INLINE_FUNCTION auto operator()(
+            TensorType const& partial_derivatives_f,
+            MappingCoord const& coord) const
+    {
+        double J = m_mapping.jacobian(coord);
+        LeviCivitaTensor<double, BasisSpatial> eps(J);
+        return tensor_mul(index<'k', 'l', 'm'>(eps), index<'l', 'm'>(partial_derivatives_f));
+    }
```

Comment:
Is it possible to be more specific about the types here? I am concerned that the errors may be a little opaque. E.g. if a user passes a contravariant object as `partial_derivatives_f` instead of a covariant object they will get an error such as:
```
Index is not compatible with tensor type
```
coming from code that they didn't write

Reply:
I am also considering that.
If I can extract the underlying types of Tensor, I can add a static_assert to check that the tensor is <cov, cov>.

Reply:
```suggestion
    template<class T, class CovBasisF>
    KOKKOS_INLINE_FUNCTION Tensor<T, get_contravariant_dims_t<CovBasisF>> operator()(
            Tensor<T, CovBasisF, CovBasisSpatial> const& partial_derivatives_f,
            MappingCoord const& coord) const
    {
        static_assert(is_covariant_vector_index_set<CovBasisF>, "Expected the derivative of f expressed in the covariant basis");
        double J = m_mapping.jacobian(coord);
        LeviCivitaTensor<double, BasisSpatial> eps(J);
        return tensor_mul(index<'k', 'l', 'm'>(eps), index<'l', 'm'>(partial_derivatives_f));
    }
```
?

Reply:
> since B may be cov or contra.

Ok. In that case this type shouldn't have `Cov` in the name